sw/device/lib/dif/dif_pwrmgr.c

```diff
@@ -4,8 +4,543 @@
+ * called.
+ */
+static void sync_slow_clock_domain(const dif_pwrmgr_t *pwrmgr) {
+  // Start sync and wait for it to finish.
+  mmio_region_write32(pwrmgr->params.base_addr, PWRMGR_CFG_CDC_SYNC_REG_OFFSET,
+                      bitfield_bit32_write(0, PWRMGR_CFG_CDC_SYNC_SYNC, true));
+  while (bitfield_bit32_read(mmio_region_read32(pwrmgr->params.base_addr,
+                                                PWRMGR_CFG_CDC_SYNC_REG_OFFSET),
+                             PWRMGR_CFG_CDC_SYNC_SYNC)) {
+  }
+}
+
+/**
+ * Checks if sources of a request type are locked.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool request_sources_is_locked(const dif_pwrmgr_t *pwrmgr,
+                                      dif_pwrmgr_req_type_t req_type) {
+  request_reg_info_t reg_info = request_reg_infos[req_type];
+  uint32_t reg_val = mmio_region_read32(pwrmgr->params.base_addr,
+                                        reg_info.write_enable_reg_offset);
+  // Locked if write enable bit is set to 0.
+  return !bitfield_bit32_read(reg_val, reg_info.write_enable_bit_index);
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_init(dif_pwrmgr_params_t params,
+                                    dif_pwrmgr_t *pwrmgr) {
+  if (pwrmgr == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  *pwrmgr = (dif_pwrmgr_t){.params = params};
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_config_result_t dif_pwrmgr_low_power_set_enabled(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_toggle_t new_state) {
+  if (pwrmgr == NULL || !is_valid_toggle(new_state)) {
+    return kDifPwrmgrConfigResultBadArg;
+  }
+
+  if (control_register_is_locked(pwrmgr)) {
+    return kDifPwrMgrConfigResultLocked;
+  }
+
+  // Read-modify-write.
+  uint32_t reg_val =
+      mmio_region_read32(pwrmgr->params.base_addr, PWRMGR_CONTROL_REG_OFFSET);
+  reg_val = bitfield_bit32_write(reg_val, PWRMGR_CONTROL_LOW_POWER_HINT,
+                                 new_state == kDifPwrmgrToggleEnabled);
+  mmio_region_write32(pwrmgr->params.base_addr, PWRMGR_CONTROL_REG_OFFSET,
+                      reg_val);
+
+  // Sync changes to slow clock domain.
+  sync_slow_clock_domain(pwrmgr);
+
+  return kDifPwrmgrConfigResultOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_low_power_get_enabled(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_toggle_t *cur_state) {
+  if (pwrmgr == NULL || cur_state == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  uint32_t reg_val =
+      mmio_region_read32(pwrmgr->params.base_addr, PWRMGR_CONTROL_REG_OFFSET);
+  *cur_state = bitfield_bit32_read(reg_val, PWRMGR_CONTROL_LOW_POWER_HINT)
+                   ? kDifPwrmgrToggleEnabled
+                   : kDifPwrmgrToggleDisabled;
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_config_result_t dif_pwrmgr_set_domain_config(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_domain_config_t config) {
+  if (pwrmgr == NULL || !is_valid_for_bitfield(config, kDomainConfigBitfield)) {
+    return kDifPwrmgrConfigResultBadArg;
```

Comment:
@lenary Can we get a ruling? I think we've pretty consistently made an exception for result_t enums where the enum variants don't need to include the token `Result`. Is this A Thing or am I making it up?

Reply:
From the [DIF style guide](https://docs.opentitan.org/sw/device/lib/dif/README/):
> dif_<ip>_<operation>_result_t is an enum representing more specific failure modes.

In this case, `Config` is not due to the name of the function but is used for all operations that modify configuration registers that can be locked. Can rename to something else but I think this is good enough.

Reply:
If `Config` is the operation, then we've said in the past that this enum member should be `kDifPwrmgrConfigBadArg` - *however* if that will conflict with other declarations, absolutely include `Result`.

Reply:
My mistake, I somehow totally misread the bullets following the section I quoted, done.